LM Format Enforcer Guided Decoding Support

noamgat opened the pull request:
FIX #3713
This PR implements the issue raised by simon-mo to add LMFE decoding support to the OpenAI server via a command line argument.

It introduces a new command line arg
`--guided-decoding-backend` which defaults to outlines (current implementation) but also support lm-format-enforcer.

Example curl command to test with:

If you start the server with
```
python -m vllm.entrypoints.openai.api_server --guided-decoding-backend lm-format-enforcer --model meta-llama/Llama-2-7b-chat-hf
```
You can then run
```
curl http://localhost:8000/v1/completions \
-H "Content-Type: application/json" \
-d '{
"model": "meta-llama/Llama-2-7b-chat-hf",
"prompt": "Please create a random person's name.",
"max_tokens": 100,
"temperature": 0,
"guided_json": {  "$schema": "http://json-schema.org/draft-04/schema#",  "description": "",  "type": "object",  "properties": {    "name": {      "type": "object",      "properties": {        "first": {          "type": "string",          "minLength": 1        },        "middle": {          "type": "string",          "minLength": 1        },        "last": {          "type": "string",          "minLength": 1        }      },      "required": [        "first",        "middle",        "last"      ]    }  },  "required": [    "name"  ]}
}'
```

And get a proper json response.

noamgat committed:
WIP: Refactoring before lm-format-enforcer integration

noamgat committed:
Integrated LM Format enforcer decoding

noamgat committed:
Merge branch 'main' into lmfe-generation

# Conflicts:
#	requirements-rocm.txt
#	vllm/config.py
#	vllm/engine/arg_utils.py

noamgat committed:
format.sh

noamgat committed:
Making pip install work

noamgat committed:
LMFE integration fixes

noamgat committed:
Ruff error

noamgat committed:
isort fix

noamgat committed:
Fixing test import

noamgat committed:
Merge branch 'main' into lmfe-generation

# Conflicts:
#	requirements-common.txt
#	requirements-rocm.txt

noamgat committed:
Refactor : introduced DecodingConfig

noamgat committed:
Added tests for get_guided_processors wrapper and LMFE implementation

noamgat committed:
format.sh

noamgat committed:
Fixing test_guided_logits_processor_black_box()

noamgat commented:
The checks caused me to realize some gaps in my original PR. They have all been addressed, this PR is 100% ready to be reviewed.

noamgat committed:
Merge branch 'main' into lmfe-generation

# Conflicts:
#	requirements-common.txt

simon-mo commented on vllm/config.py:
```diff
@@ -1000,6 +998,20 @@ def _get_and_verify_max_len(
+@dataclass
+class DecodingConfig:
+    """Dataclass which contains the decoding strategy of the engine"""
+    guided_decoding_backend: str = 'outlines'
+    """Which guided decoding algo to use. 'outlines' / 'lm-format-enforcer'"""
```
```suggestion

    # Which guided decoding algo to use. 'outlines' / 'lm-format-enforcer'
    guided_decoding_backend: str = 'outlines'
```

simon-mo commented on vllm/config.py:
```diff
@@ -73,26 +73,24 @@ class ModelConfig:
+                 quantization_param_path: Optional[str] = None,
+                 enforce_eager: bool = False,
+                 max_context_len_to_capture: Optional[int] = None,
+                 max_logprobs: int = 5) -> None:
```
```suggestion
                 max_logprobs: int = 5,) -> None:
```

doesn't look like we added anything? add the trailing comma back and return formatter.

simon-mo reviewed (approved):
This is pretty good! However, it would be even better if you can help separate out the tests in tests/test_entrypoints/test_openai_server that tests guided decoding into a separate file and we can create a fixture over the server (perhaps different ports) to run end to end test cases. In that case we would have a lot of confidence in recommending both solutions.

noamgat committed:
Update vllm/config.py based on PR review

Co-authored-by: Simon Mo <simon.mo@hey.com>

noamgat committed:
Update vllm/config.py based on PR review

Co-authored-by: Simon Mo <simon.mo@hey.com>

noamgat committed:
format.sh

noamgat committed:
Added the option to override guided_decoding_backend on a per-request basis

noamgat committed:
Modifying outlines logits processor to return the tokenizer to its normal state after using it

noamgat committed:
Ruff fixes

noamgat committed:
yapf fixes

noamgat committed:
Test fixes

noamgat committed:
Ruff fixes

noamgat committed:
yapf

noamgat committed:
isort

noamgat commented:
In order to make the tests simpler, I ended up adding a per-request option as well (same idea `guided_decoding_backend` alongside `guided_json` etc), and run the tests for both outlines and LMFE.
It required a small change to the outlines implementation, the current one modified the input tokenizer in a breaking fashion, I modified it to work on a copy of it (and added lru_cache to avoid a performance hit)

jamestwhedbee commented:
Anything else you think is blocking a merge @simon-mo?

noamgat committed:
Merge branch 'main' into lmfe-generation

noamgat commented:
The failing test is due to a temporary networking issue during the run.

```
E       socket.gaierror: [Errno -3] Temporary failure in name resolution
 ```

noamgat committed:
Doc change for CI retrigger

noamgat commented:
@simon-mo any update on the review?

simon-mo committed:
Merge branch 'main' of github.com:vllm-project/vllm into lmfe-generation
